### ISSUE TITLE
How to use embeddings-for-trees output?

### ISSUE THREAD

**Issue by SamraMehboob** [JetBrains-Research/embeddings-for-trees#15]
Hi,
I'm using this lib to generate embeddings so that I can use them as input to ML model. But output is in form of .ckpt files and I'm finding difficulty to use it. 
I tried using wandb.restore() but no success.
I'm a beginner, can you please provide me help regarding output format. I was expecting some form of numerical embedding for ASTs.
Thank you.

**Comment by SamraMehboob** [comment]
Hi
Can you please help me with this.
Thanks.

**Comment by SpirinEgor** [comment]
Hi!

`.ckpt` files correspond to model checkpoints with weights.
If you want to extract embedding you need to write some code :)
1. First of all, you need to load checkpoint. See this [example](https://github.com/JetBrains-Research/embeddings-for-trees/blob/master/embeddings_for_trees/treelstm.py#L44).
2. Since you don't need a decoder, extract embedding and encoder modules:
```
node_emb = model._TreeLSTM2Seq__embedding
tree_enc = model._TreeLSTM2Seq__encoder
```
Sorry for not very convenient code, I will look forward to provide more accurate interfaces.
3. Retrieve trees embeddings
```
batched_trees.ndata["x"] = self.__embedding(batched_trees)
encoded_nodes = self.__encoder(batched_trees)
batched_encoded_nodes, mask = cut_into_segments(
    encoded_nodes, batched_trees.batch_num_nodes(), False
)
```
4. Now you have embeddings of all nodes in the tree, you can aggregate them by mean for example.

**Comment by SamraMehboob** [comment]
Thanks a lot for detailed explanation. I'll try above steps and will let you know.

**Comment by SamraMehboob** [comment]
Sorry to disturb you again.

 **batched_encoded_nodes, mask = cut_into_segments(
        encoded_nodes, batched_trees.batch_num_nodes(), False
        )**
        Where to put this code patch?

**Comment by SpirinEgor** [comment]
Write after extracting encoded_nodes.
To speed up computation, all trees in the batch are collated into a single tree. This function cuts this tree back, so you will have encoded nodes for each tree in batch and mask to properly slice it.

**Comment by SamraMehboob** [comment]
**cut_into_segments** function is undefined. Do I  need to write this script on own or provided by some library?

**Comment by SpirinEgor** [comment]
You can find this function in the [`commode-utils`](https://github.com/SpirinEgor/commode-utils/blob/master/commode_utils/training.py#L13) package. You can also install it with `pip`.

**Comment by SamraMehboob** [comment]
Thank you.

**Comment by SamraMehboob** [comment]
Hi,
I created vector embedding for Java files.
Basically, I want to train a model for program repair task, such that it takes input-embeddings of buggy line and learn to generate output for fixed line.
I gave input buggy line and created embeddings as shown below. But I'm confused, how to convert these vectors in source code. As which value in vectors corresponding to which token?

Sorry for this silly question, but if you can provide some pointers it will be really helpful.
Thanks.

![image](https://user-images.githubusercontent.com/60770335/153218162-75ebafdf-8492-428e-bed7-09142aff76c8.png)

**Comment by SpirinEgor** [comment]
Wow, it is quite surprising for me that this even works :)
I use PyTorch as a framework to train the model and you use TensorFlow functions to operate with torch tensors...

Answering your questions:
- You may see that `encoded_nodes` passed into the decoder module. This module generates a sequence of `output_length` size. `output_logits` is a matrix with the shape `[sequence length; batch size; vocab size]`.  So, if you want to generate some code, you should: (1) create a vocabulary with possible tokens for generating (label vocabulary), (2) pass the required output length to the forward method.
- `batched_encoded_nodes` shape should be something like `[batch size; size of max tree in the batch]`. The first vector corresponds to the root vector, the following vectors correspond to the tree with respect to their definition.

**Comment by SamraMehboob** [comment]
Alright thanks for help. I'll try with above points.

**Comment by SpirinEgor** [comment]
I close it due to inactivity, but feel free to reopen or create new issues in case of any questions!
